File: apps/ios/Exponential/UI/Session/EditedFilesCard.swift
import ExpCore
import ExpUI
import SwiftUI

/// EXP-916 — the ONE "edited files" card, ×4 (web `@exp/ui` `EditedFilesCard`,
/// desktop `steer::feed`, Android `EditedFilesCard`).
///
/// A run of consecutive edit calls is ONE card in the transcript
/// (`AgentFeedRow.edits`), not a tool row per call and not a second diff view:
/// its title is what the run DID (`3 files edited`) and each row is a real
/// `DiffFileCard` — the same per-file unit the Changes face draws — stacked
/// FLUSH inside the card and separated by hairlines.
///
/// What the card owns:
/// - the open set: ONE optional set of paths. `nil` = follow the live row —
///   while the run's last call is the transcript's live row THAT row is open by
///   itself, and it folds again the moment the transcript moves on
///   (`liveIndex` → nil). A tap replaces the default with the READER's own set,
///   seeded from what is effectively open, so the live row can be collapsed
///   like any other — and that set survives the settle, because their opens are
///   theirs. Every open body is capped at the contract's inline height. A tap
///   toggles a row IN PLACE — the card never navigates to the Changes face, and
///   there is no "Revert" and no "Show changes".
/// - the publisher's cut: `truncatedLines` under the rows, the same note a file
///   list carries, so a card never silently shows a short diff (EXP-786).
/// - the fold: `diffUi.cardPreviewFiles` rows, then `{n} more` / `Show less`.
///   A live row past the preview clamps the card open so the reader can see
///   what the agent is writing.
struct EditedFilesCard: View {
    /// The card's members, in publish order (the projection's row payload).
    let items: [AgentFeedItem]
    /// The transcript's live tool row (`AgentFeed.liveToolRowId`), or nil.
    var liveItemId: Int?

    /// The reader's OWN open set, or nil while the card still follows the live
    /// row. A tap seeds it from the effective set and owns it from then on.
    @State private var readerOpen: Set<String>?
    /// The reader asked for every row.
    @State private var showAll = false
    /// The per-card parse cache — the patches are re-parsed only when the card
    /// actually changed, never once per frame.
    @State private var memo = EditCardMemo()

    private var card: EditCard.View { memo.card(items, liveItemId: liveItemId) }

    var body: some View {
        let view = card
        let open = openPaths(view)
        VStack(alignment: .leading, spacing: 0) {
            title(view)
            ForEach(shown(view)) { row in
                GlassDivider()
                fileRow(row, open: open.contains(row.path))
            }
            if view.truncatedLines > 0 {
                // EXP-786: what the PUBLISHER cut off the members' patches —
                // the same note a file list carries.
                GlassDivider()
                truncationNote(view.truncatedLines)
            }
            if let fold = foldLabel(view) {
                GlassDivider()
                foldRow(fold)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .glassCard()
        .accessibilityElement(children: .contain)
        .accessibilityIdentifier("edited-files-card")
    }

    // MARK: - Pieces

    private func title(_ view: EditCard.View) -> some View {
        HStack(spacing: 8) {
            AppIcon(AppIcons.codingDiff, size: 11)
                .foregroundStyle(.white.opacity(TextOpacity.tertiary))
            Text(view.title)
                .font(.caption.weight(.medium))
                .foregroundStyle(.white)
            Spacer(minLength: 0)
        }
        .padding(.horizontal, 12)
        .padding(.vertical, 8)
    }

    private func fileRow(_ row: EditCard.Row, open: Bool) -> some View {
        DiffFileCard(
            file: row.file ?? Diff.File(path: row.path),
            expanded: open,
            compact: true,
            flush: true,
            state: row.state,
            // EVERY open patch is capped, not just the live one: a long
            // file a reader taps open pushes the conversation off screen
            // exactly as the growing live one would.
            maxBodyHeight: Self.inlineDiffMaxHeight,
            onToggle: { toggle(row.path) }
        )
    }

    /// The publisher's own dropped-line count, ×4 (`AgentFeed`'s note).
    private func truncationNote(_ lines: Int) -> some View {
        Text(AgentFeed.diffTruncationNote(lines))
            .font(.caption2)
            .foregroundStyle(.white.opacity(TextOpacity.tertiary))
            .padding(.horizontal, 12)
            .padding(.vertical, 6)
            .frame(maxWidth: .infinity, alignment: .leading)
            .accessibilityIdentifier("edited-files-truncation-note")
    }

    private func foldRow(_ label: String) -> some View {
        Button {
            showAll.toggle()
        } label: {
            HStack(spacing: 6) {
                Text(label)
                    .font(.caption2)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                Spacer(minLength: 0)
            }
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityIdentifier("edited-files-card-fold")
    }

    // MARK: - The open set

    /// What is effectively open: the reader's set once they have touched the
    /// card, else the live row alone (and nothing at all once it is gone).
    private func openPaths(_ view: EditCard.View) -> Set<String> {
        if let readerOpen { return readerOpen }
        guard let live = view.liveIndex, view.rows.indices.contains(live)
        else { return [] }
        return [view.rows[live].path]
    }

    /// A tap toggles the path in a COPY of the EFFECTIVE set — so the live row
    /// can be collapsed like any other, and the reader's opens outlive it.
    private func toggle(_ path: String) {
        var next = openPaths(card)
        if next.contains(path) {
            next.remove(path)
        } else {
            next.insert(path)
        }
        readerOpen = next
    }

    // MARK: - Rules

    /// Web's `max-h-72` as the contract states it — the ceiling on the ONE
    /// row a live card opens by itself.
    private static let inlineDiffMaxHeight =
        CGFloat(DomainContract.diffUiInlineDiffMaxHeight)

    /// A live row past the preview clamps the card open: the reader must see
    /// the file the agent is writing right now.
    private func clamped(_ view: EditCard.View) -> Bool {
        guard let live = view.liveIndex else { return false }
        return live >= EditCard.preview
    }

    private func shown(_ view: EditCard.View) -> [EditCard.Row] {
        guard !showAll, !clamped(view), view.rows.count > EditCard.preview
        else { return view.rows }
        return Array(view.rows.prefix(EditCard.preview))
    }

    /// `{n} more` / `Show less`, or nil when there is nothing folded away (and
    /// while a live row clamps the card open — that fold is not the reader's).
    private func foldLabel(_ view: EditCard.View) -> String? {
        guard !clamped(view) else { return nil }
        if showAll { return DomainContract.diffUiShowLess }
        return EditCard.moreLabel(view.rows.count)
    }
}

/// EXP-916 — the per-card parse cache.
///
/// `EditCard.card` parses every member's patch, and a SwiftUI body runs on any
/// state change in the transcript (every streamed token of a live run). The
/// card is identified by its first and last member plus the total patch length:
/// a growing run changes its last id, a `tool_update` that lands a patch
/// changes the length, and nothing else can change what the card renders.
///
/// Deliberately NOT `@Observable`: the cache is written DURING a body pass, and
/// an observed write there would invalidate the view that just read it.
@MainActor
final class EditCardMemo {
    private var key: String?
    private var cached: EditCard.View?

    func card(_ items: [AgentFeedItem], liveItemId: Int?) -> EditCard.View {
        var length = 0
        for item in items {
            if case let .tool(_, _, _, _, _, _, settled, failed, diff, _, _) = item {
                length += diff?.utf8.count ?? 0
                if settled { length += 1 }
                if failed { length += 2 }
            }
        }
        let key = [
            String(items.first?.id ?? -1),
            String(items.last?.id ?? -1),
            String(items.count),
            String(length),
            String(liveItemId ?? -1),
        ].joined(separator: "-")
        if key == self.key, let cached { return cached }
        let made = EditCard.card(items, liveItemId: liveItemId)
        self.key = key
        cached = made
        return made
    }
}

File: apps/android/app/src/main/java/com/exponential/app/ui/session/EditedFilesCard.kt
package com.exponential.app.ui.session

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.key
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.exponential.app.domain.AgentFeedItem
import com.exponential.app.domain.Diff
import com.exponential.app.domain.DomainContract
import com.exponential.app.domain.EditCard
import com.exponential.app.domain.diffTruncationNote
import com.exponential.app.ui.icons.ExpIcons
import com.exponential.app.ui.issue.DiffCardState
import com.exponential.app.ui.issue.DiffFileCard
import com.exponential.app.ui.theme.GlassTokens
import com.exponential.app.ui.theme.TextEmphasis
import com.exponential.app.ui.theme.glassCard

// EXP-916 — the EDITED-FILES card: ONE card per run of consecutive edit calls
// ([com.exponential.app.domain.AgentFeedRow.Edits]), one FLUSH [DiffFileCard]
// per path inside it. Byte-identical ×4 (web `@exp/ui` `EditedFilesCard`,
// desktop `ui::edit_card`, iOS `EditedFilesCard.swift`): the title, the fold
// label and the preview count all come from `EditCard` / the contract.
//
// The card is TERMINAL: a tap opens a file's patch in place, and nothing on it
// ever navigates to the Changes face. No "Revert", no "Show changes".

/** The inline patch's ceiling — the contract's number, `max-h-72` on the web. */
private val InlineDiffMaxHeight = DomainContract.diffUiInlineDiffMaxHeight.dp

@Composable
fun EditedFilesCard(
    items: List<AgentFeedItem.Tool>,
    /** The id [com.exponential.app.domain.liveToolRowId] names, or null. */
    liveItemId: Long?,
    modifier: Modifier = Modifier,
) {
    if (items.isEmpty()) return
    val rowId = items.first().id
    val lastId = items.last().id
    // Parsing every member's patch is the expensive part, so the card is keyed
    // on what can actually change it: which members it covers, how much patch
    // has landed, how many have settled, and which row is live. Never a
    // per-recomposition re-parse.
    val totalDiffLength = items.sumOf { it.diff?.length ?: 0 }
    val settledCount = items.count { it.settled }
    val failedCount = items.count { it.failed }
    val card = remember(rowId, lastId, totalDiffLength, settledCount, failedCount, liveItemId) {
        EditCard.editCard(items, liveItemId)
    }
    val preview = DomainContract.diffUiCardPreviewFiles
    val liveIndex = card.liveIndex
    // The fold only ever hides rows the reader is not being shown live.
    var showAll by remember(rowId) { mutableStateOf(false) }
    val clamped = liveIndex != null && liveIndex >= preview
    val open = showAll || clamped
    val shown = if (open) card.rows else card.rows.take(preview)
    val more = EditCard.editCardMoreLabel(card.rows.size)

    // The card OWNS which files are open, as ONE nullable set. `null` = follow
    // the live row: while the card's last member is the LIVE tool row that row
    // is open by itself, so the reader watches the edit land, and the card
    // folds itself the moment it settles. A tap replaces it with the reader's
    // OWN set — seeded from what is effectively open, so the live row can be
    // collapsed like any other — and that set survives the settle, because
    // their opens are theirs.
    val livePath = liveIndex?.let { card.rows.getOrNull(it)?.path }
    var reader by remember(rowId) { mutableStateOf<Set<String>?>(null) }
    val openPaths = reader ?: if (livePath == null) emptySet() else setOf(livePath)

    Column(
        modifier = modifier.fillMaxWidth().glassCard().testTag("edited-files-card"),
    ) {
        Row(
            modifier = Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 8.dp),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            Icon(
                ExpIcons.codingDiff,
                contentDescription = null,
                modifier = Modifier.size(12.dp),
                tint = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
            )
            Spacer(Modifier.width(8.dp))
            Text(
                card.title,
                style = MaterialTheme.typography.labelSmall,
                fontWeight = FontWeight.Medium,
                color = MaterialTheme.colorScheme.onSurface,
            )
        }
        shown.forEach { row ->
            key(row.path) {
                HorizontalDivider(thickness = GlassTokens.Hairline, color = GlassTokens.StrokeRow)
                val expanded = row.path in openPaths
                val body: @Composable () -> Unit = {
                    DiffFileCard(
                        file = row.file ?: Diff.File(path = row.path),
                        expanded = expanded,
                        onToggle = {
                            reader = if (expanded) openPaths - row.path else openPaths + row.path
                        },
                        compact = true,
                        flush = true,
                        state = when (row.state) {
                            EditCard.RowState.READY -> DiffCardState.Ready
                            EditCard.RowState.PENDING -> DiffCardState.Pending
                            EditCard.RowState.DONE -> DiffCardState.Done
                            EditCard.RowState.FAILED -> DiffCardState.Failed
                        },
                    )
                }
                // An open patch — the live one grows while the call runs —
                // gets a bounded scroll box instead of pushing the
                // conversation off screen.
                if (expanded) {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .heightIn(max = InlineDiffMaxHeight)
                            .verticalScroll(rememberScrollState()),
                    ) { body() }
                } else {
                    body()
                }
            }
        }
        // EXP-786: what the PUBLISHER cut off the members' patches — the same
        // note a file list carries, so a card never silently shows a short
        // diff.
        if (card.truncatedLines > 0) {
            HorizontalDivider(thickness = GlassTokens.Hairline, color = GlassTokens.StrokeRow)
            Text(
                diffTruncationNote(card.truncatedLines),
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Tertiary),
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag("edited-files-truncation-note")
                    .padding(horizontal = 12.dp, vertical = 5.dp),
            )
        }
        // `N more` / `Show less` — the contract's words. Clamped open while the
        // live row sits past the preview: folding it away would hide the one
        // thing the reader is watching.
        if (more != null && !clamped) {
            HorizontalDivider(thickness = GlassTokens.Hairline, color = GlassTokens.StrokeRow)
            Text(
                if (showAll) DomainContract.diffUiShowLess else more,
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary),
                modifier = Modifier
                    .fillMaxWidth()
                    .testTag("edited-files-more")
                    .clickable { showAll = !showAll }
                    .padding(horizontal = 12.dp, vertical = 9.dp),
            )
        }
    }
}
